File: src/Client.Desktop/View/Window.cs
using Client.Domain;
using System.Collections.Generic;
using System;
using System.Linq;
using Microsoft.Xna.Framework;
using Client.Resources;
using Microsoft.Xna.Framework.Graphics;
using Client.Desktop.Runtime;
using System.Diagnostics;

namespace Client.View
{
    /// <summary>
    /// Contains all entities which need to be visible in Game window and converts them
    /// to textures.
    /// </summary>
    public sealed class Window : IGameComponent,
                                 IBatchDrawer,
                                 ITextureConsumer
    {
        private readonly Dictionary<GeoPoint, LocationType> mapPoints = new Dictionary<GeoPoint, LocationType>();

        // defines single texture generator strategy
        // area: small array 3*3, where central point defines the point which need th have
        // generated texture, and the rest of array items defines neighbers with indexes as constatns above.
        // if strategy can't generate texture, returns defaultTexture.
        private delegate TextureHolder Generator(LocationType[] area, TextureHolder defaultTexture);

        private WaterTextures water;
        private TextureHolder city;
        private List<Generator> functionStrategies = new List<Generator>();
        private ITileStrategy[] strategies = new ITileStrategy[0];
        private ITileStrategy fallbackStrategy;

        public event EventHandler<EventArgs> DrawOrderChanged;
        public event EventHandler<EventArgs> VisibleChanged;

        public int DrawOrder => 0;

        public bool Visible => true;

        public void Initialize(WaterTextures water, TextureHolder city, ITileStrategy fallbackStrategy, params ITileStrategy[] strategies)
        {
            this.water = water;
            this.city = city;

            functionStrategies.Add(CoastWithLandToTheWest);
            functionStrategies.Add(CoastWithLandToTheEast);
            functionStrategies.Add(CoastWithLandToTheNorthEast);
            functionStrategies.Add(CoastWithLandToTheNorthWest);
            functionStrategies.Add(CoastWithLandToTheSouthEast);
            functionStrategies.Add(CoastWithLandToTheSouthWest);

            this.strategies = strategies;
            this.fallbackStrategy = fallbackStrategy;
        }

        public void AddIsland(IslandEntity island)
        {
            foreach (var item in island.GeneratePoints().ToArray())
            {
                var point = new GeoPoint(item.X, item.Y);
                mapPoints.Add(point, LocationType.Ground);
            }
        }

        /// <summary>
        /// Defines a city on the map.
        /// 
        /// Needs to be invoked later then <see cref="AddIsland(IslandEntity)"/>
        /// </summary>
        /// <param name="entity"></param>
        public void AddCity(CityEntity entity)
        {
            var point = new GeoPoint(entity.X, entity.Y);
            mapPoints[point] = LocationType.City;
        }

        public void Include(LandUnitEntity entity)
        {
            var point = new GeoPoint(entity.X, entity.Y);
            mapPoints[point] = LocationType.LandUnit;
        }

        /// <summary>
        /// Generates list of textures for given square, when left bottom corner is provided.
        /// </summary>
        /// <param name="lbcx"></param>
        /// <param name="lbcy"></param>
        /// <param name="width"></param>
        /// <param name="height"></param>
        /// <returns>Sequence contains textures, starting from left bottom corner and returning rows first.</returns>
        public IEnumerable<PointContext> GetWindow(int lbcx, int lbcy, int width, int height)
        {
            for (int dy = 0; dy < height; dy++)
                for (int dx = 0; dx < width; dx++)
                {
                    var geox = lbcx + dx;
                    var geoy = lbcy + dy;

                    var area = new LocationType[9];
                    var centerOfArea = new GeoPoint(geox, geoy);
                    var relativeCenterOfArea = new GeoPoint(dx, dy);

                    for (int i = 0; i < 9; i++) area[i] = LocationType.Water;

                    area[Directions.NeighborTopLeft] = mapPoints.ContainsKey(centerOfArea.TopLeft()) ? mapPoints[centerOfArea.TopLeft()] : LocationType.Water;
                    area[Directions.NeighborNorth] = mapPoints.ContainsKey(centerOfArea.Top()) ? mapPoints[centerOfArea.Top()] : LocationType.Water;
                    area[Directions.NeighborTopRight] = mapPoints.ContainsKey(centerOfArea.TopRight()) ? mapPoints[centerOfArea.TopRight()] : LocationType.Water;
                    area[Directions.NeighborWest] = mapPoints.ContainsKey(centerOfArea.Left()) ? mapPoints[centerOfArea.Left()] : LocationType.Water;
                    area[Directions.NeighborThis] = mapPoints.ContainsKey(centerOfArea) ? mapPoints[centerOfArea] : LocationType.Water;
                    area[Directions.NeighborEast] = mapPoints.ContainsKey(centerOfArea.Right()) ? mapPoints[centerOfArea.Right()] : LocationType.Water;
                    area[Directions.NeighborDownLeft] = mapPoints.ContainsKey(centerOfArea.DownLeft()) ? mapPoints[centerOfArea.DownLeft()] : LocationType.Water;
                    area[Directions.NeighborSouth] = mapPoints.ContainsKey(centerOfArea.Down()) ? mapPoints[centerOfArea.Down()] : LocationType.Water;
                    area[Directions.NeighborDownRight] = mapPoints.ContainsKey(centerOfArea.DownRight()) ? mapPoints[centerOfArea.DownRight()] : LocationType.Water;

                    var centerTexture = new TextureHolder();
                    var handled = false;
                    foreach (var strategy in functionStrategies)
                    {
                        var texture = strategy(area, centerTexture);
                        if (texture == centerTexture) continue;

                        centerTexture = texture;
                        handled = true;
                        break;
                    }

                    if (!handled)
                    {
                        foreach (var strategy in strategies)
                        {
                            if (!strategy.CanExecute(area)) continue;

                            centerTexture = strategy.Execute(area);
                            handled = true;
                            break;
                        }
                    }

                    if (!handled) centerTexture = fallbackStrategy.Execute(area);

                    yield return new PointContext(relativeCenterOfArea, centerTexture);
                }
        }

        private TextureHolder CoastWithLandToTheWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheWest;
        }

        private TextureHolder CoastWithLandToTheEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheEast;
        }

        private TextureHolder CoastWithLandToTheNorthEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheNorthEast;
        }

        private TextureHolder CoastWithLandToTheNorthWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheNorthWest;
        }

        private TextureHolder CoastWithLandToTheSouthEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheSouthEast;
        }

        private TextureHolder CoastWithLandToTheSouthWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheSouthWest;
        }

        public void Initialize()
        {
            // temporar variables to keep sample textures for demo purposes.
            var island = IslandEntityGenerator.Random(new GeoPoint(20, 20));

            this.AddIsland(island);
            this.AddCity(new CityEntity(20, 20));
            this.Include(new LandUnitEntity(21, 20));
        }

        Texture2D terrainTexture;
        Texture2D desertRatsTextures;

        public void OnLoaded(Texture2D texture, TextureItem item)
        {
            if (item == TextureItem.TERRAIN) terrainTexture = texture;
            if (item == TextureItem.DESERT_RATES) desertRatsTextures = texture;
        }

        public void LoadFinished()
        {
            var waterTextures = new WaterTextures(terrainTexture);
            var cityTexture = new TextureHolder(terrainTexture, new Rectangle(7 * Config.SpriteSize, 9 * Config.SpriteSize, Config.SpriteSize, Config.SpriteSize));
            var groundTexture = new TextureHolder(terrainTexture, new Rectangle(0 * Config.SpriteSize, 0, Config.SpriteSize, Config.SpriteSize));
            var landUnitTexture = new TextureHolder(desertRatsTextures, new Rectangle(1 + 0 * Config.SpriteSize, 1 + 0, Config.SpriteSize, Config.SpriteSize));

            Initialize(waterTextures, cityTexture, new DefaultStrategy(waterTextures.Sea),
                new CoastWithLandToTheNorthStrategy(waterTextures.CoastWithLandToTheNorth),
                new CoastWithLandToTheSouthStrategy(waterTextures.CoastWithLandToTheSouth),
                new GroundStrategy(groundTexture),
                new LandUnitStrategy(landUnitTexture),
                new CityStrategy(cityTexture));
        }

        public IntendedMapCentre IntendedMapCentre { get; set; }

        public void OnDraw(SpriteBatch spriteBatch)
        {
            // get screen center
            // var offsetx = graphics.GraphicsDevice.Viewport.Width / 2;
            // var offsety = graphics.GraphicsDevice.Viewport.Height / 2;

            var x = IntendedMapCentre.X;
            var y = IntendedMapCentre.Y;

            // display sample island
            var points = this.GetWindow(0 + x, 0 + y, 30, 30);

            foreach (var it in points)
            {
                var position = new Vector2(
                    it.GeoPoint.X * Config.SpriteSize,
                    it.GeoPoint.Y * Config.SpriteSize);
                spriteBatch.Draw(position, it.Texture);
            }
        }



        /// <summary>
        /// Holds together geopoint and texture used for that point to draw it.
        /// </summary>
        public struct PointContext
        {
            public readonly GeoPoint GeoPoint;
            public readonly TextureHolder Texture;

            public PointContext(GeoPoint geopoint, TextureHolder texture)
            {
                GeoPoint = geopoint;
                Texture = texture;
            }
        }
    }
}

File: core/src/com/mygdx/game/view/Window.java
package com.mygdx.game.view;

import java.util.Set;
import java.util.function.Supplier;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.mygdx.game.Config;
import com.mygdx.game.domain.CityEntity;
import com.mygdx.game.domain.GeoPoint;
import com.mygdx.game.domain.IntendedMapCentre;
import com.mygdx.game.domain.IslandEntity;
import com.mygdx.game.domain.IslandEntityExtensions;
import com.mygdx.game.domain.IslandEntityGenerator;
import com.mygdx.game.domain.LandUnitEntity;
import com.mygdx.game.extensions.SpriteBatchUtils;
import com.mygdx.game.resources.SelectionTexture;
import com.mygdx.game.resources.TextureItem;
import com.mygdx.game.resources.WaterTextures;
import com.mygdx.game.runtime.IBatchDrawer;

import io.vavr.collection.HashMap;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.collection.Seq;

/**
 * Contains all entities which need to be visible in Game window and converts them
 * to textures.
 */
@Singleton
public final class Window {
    private Map<GeoPoint, LocationType> mapPoints = HashMap.<GeoPoint, LocationType>empty();

    private WaterTextures water;
    private List<ITileStrategy> mergeStrategies = List.empty();
    private List<ITileStrategy> strategies;

    @Inject
    private Set<IBatchDrawer> batchDrawers;

    @Inject
    public Window(WaterTextures water, Set<ITileStrategy> strategies) {
        this.water = water;

        mergeStrategies = mergeStrategies
            .append(new Water(water))
            .append(new Ground(water))
            .append(new CoastWithLandToTheEast(water))
            .append(new CoastWithLandToTheNorth(water))
            .append(new CoastWithLandToTheWest(water))
            .append(new CoastWithLandToTheNorthEast(water))
            .append(new CoastWithLandToTheSouthWest(water))
            .append(new CoastWithLandToTheSouthStrategy(water))
            .append(new WaterWithLandSouthEast(water))
            .append(new WaterWithLandNorthEast(water))
            .append(new WaterWithLandNorthSouthEast(water))
            ;

        this.strategies = List.ofAll(strategies);
    }

    public void addIsland(IslandEntity island) {
        for (var item : IslandEntityExtensions.generatePoints(island)) {
            var point = new GeoPoint(item.x, item.y);
            mapPoints = mapPoints.put(point, LocationType.Ground);
        }
    }

    /**
     * Needs to be invoked later then <see cref="AddIsland(IslandEntity)"/>.
     * @param entity
     */
    public void addCity(CityEntity entity)
    {
        var point = new GeoPoint(entity.getX(), entity.getY());
        mapPoints.put(point, LocationType.City);
    }

    /**
     * 
     * @param entity
     */
    public void include(LandUnitEntity entity)
    {
        var point = new GeoPoint(entity.getX(), entity.getY());
        mapPoints.put(point, LocationType.LandUnit);
    }

    /**
     * Generates list of textures for given square, when left bottom corner is provided.
     * @return Sequence contains textures, starting from left bottom corner and returning rows first
     */
    public Seq<PointContext> getWindow(int lbcx, int lbcy, int width, int height) {
        var result = List.<PointContext>empty();
        for (int dy = 0; dy < height; dy++) 
            for (int dx = 0; dx < width; dx++) {
                var geox = lbcx + dx;
                var geoy = lbcy + dy;

                var area = new LocationType[9];
                var centerOfArea = new GeoPoint(geox, geoy);
                var relativeCenterOfArea = new GeoPoint(dx, dy);

                for (int i = 0; i < 9; i++) area[i] = LocationType.Water;

                area[Directions.NeighborTopLeft] = mapPoints.get(centerOfArea.topLeft()).getOrElse(LocationType.Water);
                area[Directions.NeighborNorth] = mapPoints.get(centerOfArea.top()).getOrElse(LocationType.Water);
                area[Directions.NeighborTopRight] = mapPoints.get(centerOfArea.topRight()).getOrElse(LocationType.Water);
                area[Directions.NeighborWest] = mapPoints.get(centerOfArea.left()).getOrElse(LocationType.Water);
                area[Directions.NeighborThis] = mapPoints.get(centerOfArea).getOrElse(LocationType.Water);
                area[Directions.NeighborEast] = mapPoints.get(centerOfArea.right()).getOrElse(LocationType.Water);
                area[Directions.NeighborDownLeft] = mapPoints.get(centerOfArea.downLeft()).getOrElse(LocationType.Water);
                area[Directions.NeighborSouth] = mapPoints.get(centerOfArea.down()).getOrElse(LocationType.Water);
                area[Directions.NeighborDownRight] = mapPoints.get(centerOfArea.downRight()).getOrElse(LocationType.Water);


                var allStrategies = strategies.appendAll(mergeStrategies);
                var handler = allStrategies.filter(it -> it.canExecute(area)).headOption();
                var centerTexture = handler.map(it -> it.execute(area)).getOrElse(water.getBlank());

                result = result.append(new PointContext(relativeCenterOfArea, centerTexture));
            }
        return result;
    }

    public void initialize()
    {
        // temporar variables to keep sample textures for demo purposes.
        var island = IslandEntityGenerator.random(new GeoPoint(5, 5));

        this.addIsland(island);
        // var island1 = new IslandEntity(new GeoPoint[] {new GeoPoint(1, 1), new GeoPoint(2, 2)});
        // this.addIsland(island1);
        this.addCity(new CityEntity(10, 10));
        this.include(new LandUnitEntity(11, 10));
    }

    Texture terrainTexture;
    Texture desertRatsTextures;

    public void onLoaded(Texture texture, TextureItem item) {
        if (item == TextureItem.TERRAIN) terrainTexture = texture;
        if (item == TextureItem.DESERT_RATES) desertRatsTextures = texture;
    }

    @Inject
    IntendedMapCentre intendedMapCentre;

    public void onDraw(SpriteBatch spriteBatch) {

        var x = intendedMapCentre.getX();
        var y = intendedMapCentre.getY();

        // display sample island
        var points = this.getWindow(0 + x, 0 + y, 20, 20);

        for (var it : points) {
            var position = new Vector2(
                it.geoPoint.x * Config.SpriteSize,
                it.geoPoint.y * Config.SpriteSize);
            if (it.texture == null) {
                assert false;
            }
            SpriteBatchUtils.draw(spriteBatch, position, it.texture);
        }

        for (var drawer : this.batchDrawers) {
            drawer.OnDraw(spriteBatch);
        }
    }

    @Inject
    SelectionTexture st;


    /**
     * Holds together geopoint and texture used for that point to draw it.
     */
    public class PointContext {
        public final GeoPoint geoPoint;
        public final TextureHolder texture;

        public PointContext(GeoPoint geopoint, TextureHolder texture) {
            this.geoPoint = geopoint;
            this.texture = texture;
        }
    }
}

// defines single texture generator strategy
// area: small array 3*3, where central point defines the point which need th have
// generated texture, and the rest of array items defines neighbers with indexes as constatns above.
// if strategy can't generate texture, returns defaultTexture.
interface MergeStrategy extends ITileStrategy {
}

final class CoastWithLandToTheEast implements MergeStrategy {
    private WaterTextures water;
    CoastWithLandToTheEast(WaterTextures water) {
        this.water = water;
    }

    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;
        return true;
    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return water.getLandEast();
    }

}

final class CoastWithLandToTheNorth implements MergeStrategy {

    private WaterTextures water;

    CoastWithLandToTheNorth(WaterTextures water) {
        this.water = water;
    }

    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] != LocationType.Ground) return false;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;
        return true;
    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return water.getLandNorth();
    }

}

final class CoastWithLandToTheWest implements MergeStrategy {

    private WaterTextures water;

    CoastWithLandToTheWest(WaterTextures water) {
        this.water = water;
    }

    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;
        return true;
    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return water.getLandWest();
    }

}

final class CoastWithLandToTheNorthEast implements MergeStrategy {

    private WaterTextures water;

    CoastWithLandToTheNorthEast(WaterTextures water) {
        this.water = water;
    }
    
    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;
        return true;
    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return water.getLandNorthEast();
    }
}

final class CoastWithLandToTheSouthWest implements MergeStrategy {

    private WaterTextures water;

    CoastWithLandToTheSouthWest(WaterTextures water) {
        this.water = water;
    }
    
    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborSouth] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;
    
        return true;
    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return water.getLandSouthWest();
    }
}

final class CoastWithLandToTheSouthStrategy implements ITileStrategy {

    private final WaterTextures textures;

    @Inject
    public CoastWithLandToTheSouthStrategy(WaterTextures waterTextures) {
        this.textures = waterTextures;
    }

    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return false;
        if (neighbors[Directions.NeighborSouth] == LocationType.Water) return false;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return false;

        return true;

    }

    public TextureHolder execute(LocationType[] neighbors) {
        return textures.getLandSouth();
    }
}

final class WaterWithLandSouthEast extends BaseGroundStrategy {

    @Inject
    public WaterWithLandSouthEast(WaterTextures waterTextures) {
        super(LocationType.Water, LocationType.Ground, LocationType.Ground, LocationType.Water,
        LocationType.Water,
        waterTextures::getLandSouthEast);
    }
}

final class WaterWithLandNorthEast extends BaseGroundStrategy {

    @Inject
    public WaterWithLandNorthEast(WaterTextures waterTextures) {
        super(LocationType.Ground, LocationType.Water, LocationType.Water, LocationType.Ground,
        LocationType.Water,
        waterTextures::getLandNorthWest);
    }
}

final class WaterWithLandNorthSouthEast extends BaseGroundStrategy {

    @Inject
    public WaterWithLandNorthSouthEast(WaterTextures waterTextures) {
        super(LocationType.Ground, LocationType.Ground, LocationType.Ground, LocationType.Water,
        LocationType.Water,
        waterTextures::getLandNorthSouthEast);
    }
}

final class Ground implements ITileStrategy {

    private final WaterTextures textures;

    @Inject
    public Ground(WaterTextures waterTextures) {
        this.textures = waterTextures;
    }

    @Override
    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborThis] != LocationType.Ground) return false;
        return true;

    }

    @Override
    public TextureHolder execute(LocationType[] neighbors) {
        return textures.getGround();
    }
}

final class Water extends BaseGroundStrategy {

    @Inject
    public Water(WaterTextures waterTextures) {
        super(LocationType.Water, LocationType.Water, LocationType.Water, LocationType.Water,
        LocationType.Water,
        waterTextures::getSea);
    }
}


abstract class BaseGroundStrategy implements ITileStrategy {
    private final LocationType north;
    private final LocationType south;
    private final LocationType east;
    private final LocationType west;
    private final LocationType self;
    private final Supplier<TextureHolder> supplier;

    @Inject
    protected BaseGroundStrategy(
        LocationType north, LocationType south, LocationType east, LocationType west,
        LocationType self,
        Supplier<TextureHolder> supplier) {
        this.north = north;
        this.south = south;
        this.east = east;
        this.west = west;
        this.self = self;
        this.supplier = supplier;
    }

    public boolean canExecute(LocationType[] neighbors) {
        if (neighbors[Directions.NeighborNorth] != north) return false;
        if (neighbors[Directions.NeighborSouth] != south) return false;
        if (neighbors[Directions.NeighborEast] != east) return false;
        if (neighbors[Directions.NeighborWest] != west) return false;
        if (neighbors[Directions.NeighborThis] != self) return false;
        return true;

    }

    public TextureHolder execute(LocationType[] neighbors) {
        return supplier.get();
    }
}